File: rust/src/actions/trading.rs
//! Trading-related actions for Polymarket.
//!
//! This module provides functions for:
//! - Getting balances (collateral and conditional tokens)
//! - Getting positions
//! - Getting trade history
//! - Getting price history
//! - Checking order scoring

use crate::client::ClobClient;
use crate::error::{PolymarketError, Result};
use crate::types::{BalanceAllowance, OrderSide, PriceHistoryEntry, TradesResponse};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// =============================================================================
// Balances
// =============================================================================

/// Response from balance endpoints
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BalanceResponse {
    /// Collateral (USDC) balance
    pub collateral: Option<BalanceAllowance>,
    /// Conditional token balances by token ID
    pub conditional: HashMap<String, BalanceAllowance>,
}

/// Get account balances.
///
/// # Arguments
///
/// * `client` - Authenticated CLOB client
/// * `token_ids` - Optional list of token IDs to get conditional balances for
/// * `include_collateral` - Whether to include USDC collateral balance
/// * `include_conditional` - Whether to include conditional token balances
///
/// # Returns
///
/// Balance information including collateral and conditional tokens.
pub async fn get_balances(
    client: &ClobClient,
    token_ids: Option<&[String]>,
    include_collateral: bool,
    include_conditional: bool,
) -> Result<BalanceResponse> {
    if !client.has_credentials() {
        return Err(PolymarketError::auth_error(
            "API credentials required for balance queries",
        ));
    }

    let mut response = BalanceResponse {
        collateral: None,
        conditional: HashMap::new(),
    };

    // Get collateral balance if requested
    if include_collateral {
        let balance = client.get_collateral_balance().await?;
        response.collateral = Some(balance);
    }

    // Get conditional balances if requested
    if include_conditional {
        if let Some(tokens) = token_ids {
            for token_id in tokens {
                match client.get_conditional_balance(token_id).await {
                    Ok(balance) => {
                        response.conditional.insert(token_id.clone(), balance);
                    }
                    Err(e) => {
                        // Log error but continue with other tokens
                        eprintln!("Failed to get balance for token {}: {}", token_id, e);
                    }
                }
            }
        }
    }

    Ok(response)
}

// =============================================================================
// Positions
// =============================================================================

/// Position information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Position {
    /// Market condition ID
    pub market: String,
    /// Token ID
    pub asset_id: String,
    /// Position size
    pub size: String,
    /// Average entry price
    pub average_price: String,
    /// Realized profit/loss
    pub realized_pnl: String,
    /// Unrealized profit/loss
    pub unrealized_pnl: String,
}

/// Get user positions.
///
/// Builds positions from trade history, calculating realized and unrealized PnL.
///
/// # Arguments
///
/// * `client` - Authenticated CLOB client
/// * `limit` - Maximum number of trades to fetch per page
/// * `max_pages` - Maximum pages to fetch
/// * `asset_ids` - Optional filter by asset IDs
/// * `include_prices` - Whether to include current prices for unrealized PnL
///
/// # Returns
///
/// List of positions with PnL calculations.
pub async fn get_positions(
    client: &ClobClient,
    limit: Option<u32>,
    max_pages: Option<u32>,
    asset_ids: Option<&[String]>,
    include_prices: bool,
) -> Result<Vec<Position>> {
    if !client.has_credentials() {
        return Err(PolymarketError::auth_error(
            "API credentials required for position queries",
        ));
    }

    let trades = client
        .get_trades(None, asset_ids, limit, max_pages)
        .await?;

    // Build positions from trades
    let mut position_map: HashMap<String, PositionAccumulator> = HashMap::new();

    for trade in &trades.data {
        let entry = position_map
            .entry(trade.token_id.clone())
            .or_insert_with(|| PositionAccumulator {
                market: trade.market_id.clone(),
                asset_id: trade.token_id.clone(),
                total_bought: 0.0,
                total_sold: 0.0,
                total_cost: 0.0,
                total_proceeds: 0.0,
            });

        let price = trade.price.parse::<f64>().unwrap_or(0.0);
        let size = trade.size.parse::<f64>().unwrap_or(0.0);

        match trade.side {
            OrderSide::Buy => {
                entry.total_bought += size;
                entry.total_cost += price * size;
            }
            OrderSide::Sell => {
                entry.total_sold += size;
                entry.total_proceeds += price * size;
            }
        }
    }

    // Convert accumulators to positions
    let mut positions = Vec::new();

    for (asset_id, acc) in position_map {
        let net_size = acc.total_bought - acc.total_sold;
        if net_size.abs() < 0.0001 {
            continue; // Skip zero positions
        }

        let avg_price = if acc.total_bought > 0.0 {
            acc.total_cost / acc.total_bought
        } else {
            0.0
        };

        let realized_pnl = acc.total_proceeds - (acc.total_sold * avg_price);

        // Get current price for unrealized PnL if requested
        let unrealized_pnl = if include_prices && net_size > 0.0 {
            match client.get_midpoint(&asset_id).await {
                Ok(mid) => {
                    let current_price = mid.parse::<f64>().unwrap_or(0.0);
                    (current_price - avg_price) * net_size
                }
                Err(_) => 0.0,
            }
        } else {
            0.0
        };

        positions.push(Position {
            market: acc.market,
            asset_id,
            size: format!("{:.6}", net_size),
            average_price: format!("{:.4}", avg_price),
            realized_pnl: format!("{:.4}", realized_pnl),
            unrealized_pnl: format!("{:.4}", unrealized_pnl),
        });
    }

    Ok(positions)
}

/// Internal accumulator for position calculations
struct PositionAccumulator {
    market: String,
    #[allow(dead_code)]
    asset_id: String,
    total_bought: f64,
    total_sold: f64,
    total_cost: f64,
    total_proceeds: f64,
}

// =============================================================================
// Trade History
// =============================================================================

/// Get trade history.
///
/// # Arguments
///
/// * `client` - Authenticated CLOB client
/// * `market_id` - Optional market filter
/// * `asset_ids` - Optional asset filter
/// * `limit` - Maximum trades per page
/// * `max_pages` - Maximum pages to fetch
///
/// # Returns
///
/// Trade history with pagination info.
pub async fn get_trade_history(
    client: &ClobClient,
    market_id: Option<&str>,
    asset_ids: Option<&[String]>,
    limit: Option<u32>,
    max_pages: Option<u32>,
) -> Result<TradesResponse> {
    if !client.has_credentials() {
        return Err(PolymarketError::auth_error(
            "API credentials required for trade history",
        ));
    }

    client.get_trades(market_id, asset_ids, limit, max_pages).await
}

// =============================================================================
// Price History
// =============================================================================

/// Get price history for a token.
///
/// # Arguments
///
/// * `client` - CLOB client (no auth required)
/// * `token_id` - The token ID
/// * `start_ts` - Start timestamp (Unix seconds)
/// * `end_ts` - End timestamp (Unix seconds)
/// * `fidelity` - Time interval fidelity in minutes (1, 5, 15, 60, etc.)
///
/// # Returns
///
/// List of price history entries.
pub async fn get_price_history(
    client: &ClobClient,
    token_id: &str,
    start_ts: Option<i64>,
    end_ts: Option<i64>,
    fidelity: Option<u32>,
) -> Result<Vec<PriceHistoryEntry>> {
    client
        .get_price_history(token_id, start_ts, end_ts, fidelity)
        .await
}

// =============================================================================
// Order Scoring
// =============================================================================

/// Check if orders are scoring (eligible for rewards).
///
/// # Arguments
///
/// * `client` - Authenticated CLOB client
/// * `order_ids` - List of order IDs to check
///
/// # Returns
///
/// Map of order ID to scoring status.
pub async fn check_order_scoring(
    client: &ClobClient,
    order_ids: &[String],
) -> Result<HashMap<String, bool>> {
    if !client.has_credentials() {
        return Err(PolymarketError::auth_error(
            "API credentials required for order scoring check",
        ));
    }

    client.check_order_scoring(order_ids).await
}

// =============================================================================
// Order Details
// =============================================================================

/// Get details for a specific order.
///
/// # Arguments
///
/// * `client` - Authenticated CLOB client
/// * `order_id` - The order ID
///
/// # Returns
///
/// Detailed order information.
pub async fn get_order_details(
    client: &ClobClient,
    order_id: &str,
) -> Result<crate::types::OpenOrder> {
    if !client.has_credentials() {
        return Err(PolymarketError::auth_error(
            "API credentials required for order details",
        ));
    }

    client.get_order(order_id).await
}

/// Get active orders for the authenticated user.
///
/// # Arguments
///
/// * `client` - Authenticated CLOB client
/// * `market_id` - Optional market filter
/// * `asset_id` - Optional asset filter
///
/// # Returns
///
/// List of active orders.
pub async fn get_active_orders(
    client: &ClobClient,
    market_id: Option<&str>,
    asset_id: Option<&str>,
) -> Result<Vec<crate::types::OpenOrder>> {
    if !client.has_credentials() {
        return Err(PolymarketError::auth_error(
            "API credentials required for active orders",
        ));
    }

    client.get_orders(market_id, asset_id).await
}

File: python/elizaos_plugin_polymarket/actions/trading.py
from collections.abc import Callable
from typing import Protocol, cast

from elizaos_plugin_polymarket.error import PolymarketError, PolymarketErrorCode
from elizaos_plugin_polymarket.providers import get_authenticated_clob_client, get_clob_client


class RuntimeProtocol(Protocol):
    def get_setting(self, key: str) -> str | None: ...


async def check_order_scoring(
    order_ids: list[str],
    runtime: RuntimeProtocol | None = None,
) -> dict[str, bool]:
    if not order_ids:
        raise PolymarketError(
            PolymarketErrorCode.INVALID_ORDER,
            "At least one order ID is required",
        )

    try:
        client = get_authenticated_clob_client(runtime)

        # Check if client has areOrdersScoring method
        fn = getattr(client, "areOrdersScoring", None)
        if callable(fn):
            response_obj = cast(Callable[[dict[str, object]], object], fn)({"orderIds": order_ids})
            if isinstance(response_obj, dict):
                return {
                    str(k): bool(v)
                    for k, v in response_obj.items()
                    if isinstance(k, str) or isinstance(k, int)
                }

            # Convert scalar response to dict if needed
            return {order_id: bool(response_obj) for order_id in order_ids}
        else:
            # Fallback: return False for all orders if method not available
            raise PolymarketError(
                PolymarketErrorCode.API_ERROR,
                "areOrdersScoring method not available in CLOB client. "
                "Please use the Polymarket API directly or update py-clob-client.",
            )

    except PolymarketError:
        raise
    except Exception as e:
        raise PolymarketError(
            PolymarketErrorCode.API_ERROR,
            f"Failed to check order scoring: {e}",
            cause=e,
        ) from e


async def get_active_orders(
    market_id: str | None = None,
    asset_id: str | None = None,
    runtime: RuntimeProtocol | None = None,
) -> list[dict[str, object]]:
    """
    Get active orders for the authenticated user.

    Args:
        market_id: Optional market condition ID filter
        asset_id: Optional asset ID filter
        runtime: Optional agent runtime for settings

    Returns:
        List of active orders

    Raises:
        PolymarketError: If fetching orders fails
    """
    try:
        client = get_authenticated_clob_client(runtime)

        # Build parameters
        params: dict[str, str] = {}
        if market_id:
            params["market"] = market_id
        if asset_id:
            params["asset_id"] = asset_id

        # Use getOpenOrders if available, otherwise fallback to get_orders
        get_open_orders = getattr(client, "getOpenOrders", None)
        get_orders = getattr(client, "get_orders", None)

        orders_obj: object
        if callable(get_open_orders):
            orders_obj = cast(Callable[[object | None], object], get_open_orders)(
                params if params else None
            )
        elif callable(get_orders):
            orders_obj = (
                cast(Callable[..., object], get_orders)(**params)
                if params
                else cast(Callable[..., object], get_orders)()
            )
        else:
            raise PolymarketError(
                PolymarketErrorCode.API_ERROR,
                "getOpenOrders or get_orders method not available in CLOB client",
            )

        if isinstance(orders_obj, dict):
            data_obj = orders_obj.get("data", [])
            if isinstance(data_obj, list):
                return [item for item in data_obj if isinstance(item, dict)]
            return []

        if isinstance(orders_obj, list):
            return [item for item in orders_obj if isinstance(item, dict)]

        return []

    except PolymarketError:
        raise
    except Exception as e:
        raise PolymarketError(
            PolymarketErrorCode.API_ERROR,
            f"Failed to fetch active orders: {e}",
            cause=e,
        ) from e


async def get_trade_history(
    market_id: str | None = None,
    asset_id: str | None = None,
    limit: int = 20,
    runtime: RuntimeProtocol | None = None,
) -> dict[str, object]:
    """
    Get trade history for the authenticated user.

    Args:
        market_id: Optional market condition ID filter
        asset_id: Optional asset ID filter
        limit: Maximum number of trades to return
        runtime: Optional agent runtime for settings

    Returns:
        Dictionary with trades list and pagination info

    Raises:
        PolymarketError: If fetching trade history fails
    """
    try:
        client = get_authenticated_clob_client(runtime)

        # Build parameters
        params: dict[str, object] = {}
        if market_id:
            params["market"] = market_id
        if asset_id:
            params["asset_id"] = asset_id

        # Use getTradesPaginated if available, otherwise getTrades
        if hasattr(client, "getTradesPaginated"):
            response = client.getTradesPaginated(params if params else None)
            if isinstance(response, dict):
                trades = response.get("trades", [])
                next_cursor = response.get("next_cursor", "")
            else:
                trades = []
                next_cursor = ""
        elif hasattr(client, "getTrades"):
            trades = client.getTrades(params if params else None)
            if isinstance(trades, dict):
                trades = trades.get("data", [])
            next_cursor = ""
        else:
            raise PolymarketError(
                PolymarketErrorCode.API_ERROR,
                "getTradesPaginated or getTrades method not available in CLOB client",
            )

        # Limit results
        if isinstance(trades, list) and limit:
            trades = trades[:limit]

        return {
            "trades": trades if isinstance(trades, list) else [],
            "next_cursor": next_cursor,
            "count": len(trades) if isinstance(trades, list) else 0,
        }

    except PolymarketError:
        raise
    except Exception as e:
        raise PolymarketError(
            PolymarketErrorCode.API_ERROR,
            f"Failed to fetch trade history: {e}",
            cause=e,
        ) from e


async def get_price_history(
    token_id: str,
    start_ts: int | None = None,
    end_ts: int | None = None,
    fidelity: int = 60,
    runtime: RuntimeProtocol | None = None,
) -> list[dict[str, object]]:
    """
    Get price history for a specific token.

    Args:
        token_id: The token ID to get price history for
        start_ts: Optional start timestamp (defaults to 24 hours ago)
        end_ts: Optional end timestamp (defaults to now)
        fidelity: Time interval in minutes (default 60)
        runtime: Optional agent runtime for settings

    Returns:
        List of price history entries with timestamp and price

    Raises:
        PolymarketError: If fetching price history fails
    """
    if not token_id:
        raise PolymarketError(
            PolymarketErrorCode.INVALID_TOKEN,
            "Token ID is required",
        )

    try:
        import time

        client = get_clob_client(runtime)

        # Set default timestamps if not provided
        now = int(time.time())
        if start_ts is None:
            start_ts = now - 86400  # 24 hours ago
        if end_ts is None:
            end_ts = now

        # Use getPricesHistory if available
        if hasattr(client, "getPricesHistory"):
            response = client.getPricesHistory(
                {
                    "market": token_id,
                    "startTs": start_ts,
                    "endTs": end_ts,
                    "fidelity": fidelity,
                }
            )

            # Convert response to list of dicts
            if isinstance(response, list):
                return [
                    {
                        "timestamp": entry.get("t", entry.get("timestamp", 0)),
                        "price": entry.get("p", entry.get("price", "0")),
                    }
                    for entry in response
                ]
            elif isinstance(response, dict):
                data = response.get("data", [])
                return [
                    {
                        "timestamp": entry.get("t", entry.get("timestamp", 0)),
                        "price": entry.get("p", entry.get("price", "0")),
                    }
                    for entry in data
                ]
            else:
                return []
        else:
            raise PolymarketError(
                PolymarketErrorCode.API_ERROR,
                "getPricesHistory method not available in CLOB client",
            )

    except PolymarketError:
        raise
    except Exception as e:
        raise PolymarketError(
            PolymarketErrorCode.API_ERROR,
            f"Failed to fetch price history: {e}",
            cause=e,
        ) from e
